File: rag/ingest.py
import os, glob, uuid, re, json
from dotenv import load_dotenv
import requests

# Load environment variables from .env
load_dotenv()

# ------------------------------
#  Config
# ------------------------------

LLM_PROVIDER = os.getenv("LLM_PROVIDER", "hf").lower()
USE_OPENAI = LLM_PROVIDER == "openai"
USE_HF = LLM_PROVIDER == "hf"

QDRANT_URL = os.getenv("QDRANT_URL", "http://localhost:6333")
COLLECTION = os.getenv("QDRANT_COLLECTION", "training_chunks")
EMBED_DIMS = int(os.getenv("EMBED_DIMS", "384"))

# Hugging Face config
HF_API_TOKEN = os.getenv("HF_API_TOKEN")
HF_EMBED_MODEL = os.getenv("HF_EMBED_MODEL", "sentence-transformers/all-MiniLM-L6-v2")

# OpenAI (optional branch)
if USE_OPENAI:
    from openai import OpenAI
    OPENAI_API_KEY = os.getenv("OPENAI_API_KEY")
    if not OPENAI_API_KEY:
        raise RuntimeError("LLM_PROVIDER=openai but OPENAI_API_KEY is missing.")
    openai_client = OpenAI(api_key=OPENAI_API_KEY)

# ------------------------------
#  Qdrant via REST
# ------------------------------

def qdrant_get_collections():
    r = requests.get(f"{QDRANT_URL}/collections")
    r.raise_for_status()
    return r.json()["result"]["collections"]

def qdrant_create_collection():
    payload = {
        "vectors": {
            "size": EMBED_DIMS,
            "distance": "Cosine"
        }
    }
    r = requests.put(f"{QDRANT_URL}/collections/{COLLECTION}", json=payload)
    r.raise_for_status()
    print(f"Created Qdrant collection: {COLLECTION}")

def ensure_collection():
    cols = [c["name"] for c in qdrant_get_collections()]
    if COLLECTION not in cols:
        qdrant_create_collection()
    else:
        print(f"Qdrant collection '{COLLECTION}' already exists.")

# ------------------------------
#  Text chunking
# ------------------------------

def chunk_text(text, max_chars=800, overlap=200):
    text = re.sub(r"\s+", " ", text).strip()
    chunks = []
    i = 0
    while i < len(text):
        j = min(i + max_chars, len(text))
        chunks.append(text[i:j])
        i = j - overlap
        if i < 0:
            i = 0
    return chunks

# ------------------------------
#  Embeddings: OpenAI or Hugging Face
# ------------------------------

def embed_batch(texts):
    # OpenAI branch (paid)
    if USE_OPENAI:
        res = openai_client.embeddings.create(
            model="text-embedding-3-small",
            input=texts
        )
        return [d.embedding for d in res.data]

    # Hugging Face Inference API branch (free tier)
    if USE_HF:
        if not HF_API_TOKEN:
            raise RuntimeError("LLM_PROVIDER=hf but HF_API_TOKEN is missing in .env")

        url = f"https://api-inference.huggingface.co/models/{HF_EMBED_MODEL}"
        headers = {"Authorization": f"Bearer {HF_API_TOKEN}"}
        # For sentence-transformers style models, sending {"inputs": [texts]} returns list-of-list embeddings
        response = requests.post(
            url,
            headers=headers,
            json={"inputs": texts}
        )
        response.raise_for_status()
        data = response.json()

        # data is usually a list of list[float], but be defensive:
        # - if single input: data may be list[float]
        # - if multiple inputs: data may be list[list[float]]
        vectors = []
        if isinstance(data, list):
            if len(data) > 0 and isinstance(data[0], list):
                # typical: list of embeddings
                vectors = data
            elif len(data) > 0 and isinstance(data[0], (int, float)):
                # single embedding
                vectors = [data]
        else:
            raise RuntimeError(f"Unexpected HF response format: {data}")

        return vectors

    raise RuntimeError("Unknown LLM_PROVIDER setting; expected 'openai' or 'hf'")

# ------------------------------
#  Qdrant upsert
# ------------------------------

def qdrant_upsert(chunks, vectors, meta):
    points = []
    for chunk, vec in zip(chunks, vectors):
        points.append({
            "id": str(uuid.uuid4()),
            "vector": vec,
            "payload": {
                **meta,
                "content": chunk
            }
        })

    payload = {"points": points}
    r = requests.put(
        f"{QDRANT_URL}/collections/{COLLECTION}/points",
        json=payload
    )
    r.raise_for_status()

# ------------------------------
#  Ingestion loop
# ------------------------------

def ingest_doc(path):
    print(f"📄 Ingesting {path}")
    # infer metadata from folder structure: data/<course>/<module>/<file>.txt
    parts = path.split(os.sep)
    meta = {
        "doc_id": os.path.splitext(os.path.basename(path))[0],
        "course_id": parts[-3] if len(parts) >= 3 else "general",
        "module_id": parts[-2] if len(parts) >= 2 else "intro",
    }

    with open(path, "r", encoding="utf-8") as f:
        text = f.read()

    chunks = chunk_text(text)
    vectors = embed_batch(chunks)
    qdrant_upsert(chunks, vectors, meta)
    print(f"✅ Ingested {len(chunks)} chunks\n")

# ------------------------------
#  MAIN
# ------------------------------

if __name__ == "__main__":
    print("🚀 Starting ingestion")
    if USE_OPENAI:
        print("🔵 Using OpenAI embeddings")
    elif USE_HF:
        print(f"🟣 Using Hugging Face embeddings via API: {HF_EMBED_MODEL}")
    else:
        print("⚠️ Unknown LLM_PROVIDER, defaulting to HF")

    ensure_collection()

    for path in glob.glob("data/**/*.txt", recursive=True):
        ingest_doc(path)

    print("🎉 DONE")

File: rag/ingest.js
import "dotenv/config";
import { InferenceClient } from "@huggingface/inference";
import axios from "axios";
import fs from "fs";
import path from "path";
import crypto from "crypto";
import * as pdfjsLib from "pdfjs-dist/legacy/build/pdf.mjs";


//
// ENVIRONMENT
//
const HF_API_TOKEN = process.env.HF_API_TOKEN;
const HF_EMBED_MODEL =
  process.env.HF_EMBED_MODEL || "sentence-transformers/all-MiniLM-L6-v2";
const EMBED_DIMS_INT = parseInt(process.env.EMBED_DIMS || "384", 10);

const QDRANT_URL = process.env.QDRANT_URL || "http://localhost:6333";
const QDRANT_COLLECTION =
  process.env.QDRANT_COLLECTION || "training_chunks";

const hfClient = new InferenceClient(HF_API_TOKEN);

//
// HELPERS
//
function chunkText(text) {
  // normalize whitespace
  const cleaned = text.replace(/\s+/g, " ").trim();

  // SIMPLE MVP: just return the whole doc as a single chunk
  if (!cleaned) return [];
  return [cleaned];
}

async function extractPagesFromPdf(filePath) {
  const data = new Uint8Array(fs.readFileSync(filePath));
  const loadingTask = pdfjsLib.getDocument({ data });
  const pdf = await loadingTask.promise;

  const pages = [];

  for (let pageNum = 1; pageNum <= pdf.numPages; pageNum++) {
    const page = await pdf.getPage(pageNum);
    const content = await page.getTextContent();

    const pageText = content.items
      .map((item) => (item.str ? item.str : ""))
      .join(" ")
      .replace(/\s+/g, " ")
      .trim();

    if (pageText) pages.push({ pageNum, text: pageText });
  }

  return pages;
}

async function extractTextFromFile(filePath) {
  const ext = path.extname(filePath).toLowerCase();

  if (ext === ".txt") {
    return fs.readFileSync(filePath, "utf8");
  }

  if (ext === ".pdf") {
    return await extractTextFromPdf(filePath);
  }

  console.warn(`Skipping unsupported file type: ${filePath}`);
  return "";
}



//
// HuggingFace embeddings
//
async function embedBatch(texts) {
  const result = await hfClient.featureExtraction({
    model: HF_EMBED_MODEL,
    inputs: texts, // HF client supports batched input
    provider: "hf-inference",
  });

  let vectors = [];

  // Case: [[float,...],[float,...]]
  if (Array.isArray(result) && Array.isArray(result[0])) {
    vectors = result;
  }

  // Case: [float,...] (single embedding)
  else if (Array.isArray(result) && typeof result[0] === "number") {
    vectors = [result];
  } else {
    throw new Error(
      "Unexpected embedding format: " + JSON.stringify(result).slice(0, 200)
    );
  }

  return vectors;
}

//
// QDRANT
//
async function ensureCollection() {
  const res = await axios.get(`${QDRANT_URL}/collections`);
  const collections = res.data.result.collections.map((c) => c.name);

  if (!collections.includes(QDRANT_COLLECTION)) {
    console.log(`Creating collection ${QDRANT_COLLECTION}...`);
    await axios.put(`${QDRANT_URL}/collections/${QDRANT_COLLECTION}`, {
      vectors: { size: EMBED_DIMS_INT, distance: "Cosine" },
    });
    console.log("Created.");
  } else {
    console.log(`Collection '${QDRANT_COLLECTION}' exists.`);
  }
}

async function upsert(chunks, vectors, meta, chunkMeta = []) {
  const points = chunks.map((chunk, i) => ({
    id: crypto.randomUUID(),
    vector: vectors[i],
    payload: { ...meta, ...(chunkMeta[i] || {}), content: chunk },
  }));

  await axios.put(
    `${QDRANT_URL}/collections/${QDRANT_COLLECTION}/points`,
    { points },
    { headers: { "Content-Type": "application/json" } }
  );
}


//
// MAIN INGEST FILE FUNCTION
//
async function ingestFile(filePath) {
  console.log(`📄 Ingesting ${filePath}`);

  const ext = path.extname(filePath).toLowerCase();

  let chunks = [];
  let chunkMeta = []; // per-chunk metadata (e.g., page_number)

  if (ext === ".pdf") {
    const pages = await extractPagesFromPdf(filePath);

    chunks = pages.map((p) => p.text);
    chunkMeta = pages.map((p) => ({ page_number: p.pageNum }));
  } else if (ext === ".txt") {
    const text = fs.readFileSync(filePath, "utf8").replace(/\s+/g, " ").trim();
    if (text) {
      chunks = [text];
      chunkMeta = [{}];
    }
  } else {
    console.warn(`Skipping unsupported file type: ${filePath}`);
    return;
  }

  if (chunks.length === 0) {
    console.warn(`No text extracted from ${filePath}, skipping.`);
    return;
  }

  // Embeddings
  const vectors = await embedBatch(chunks);

  console.log(`Chunks: ${chunks.length}`);
  console.log(
    `First embedding length: ${vectors[0]?.length}, expected: ${EMBED_DIMS_INT}`
  );

  // Metadata inferred from data/<course>/<module>/<file>
  const relParts = path.relative("data", filePath).split(path.sep);
  const meta = {
    doc_id: path.basename(filePath, ext),
    course_id: relParts[0] || "general",
    module_id: relParts[1] || "intro",
    file_type: ext.slice(1), // "pdf" or "txt"
    source_path: relParts.join("/"),
  };

  // Upsert into Qdrant (upsert must merge chunkMeta[i] into payload)
  await upsert(chunks, vectors, meta, chunkMeta);

  console.log(`✅ Ingested ${chunks.length} chunks\n`);
}


//
// RECURSIVE FIND
//
function findDocs(dir) {
  let out = [];
  for (const f of fs.readdirSync(dir)) {
    const full = path.join(dir, f);
    const stat = fs.statSync(full);
    if (stat.isDirectory()) {
      out = out.concat(findDocs(full));
    } else if (stat.isFile()) {
      const ext = path.extname(full).toLowerCase();
      if (ext === ".txt" || ext === ".pdf") {
        out.push(full);
      }
    }
  }
  return out;
}

//
// RUN
//
(async () => {
  console.log("🚀 Starting ingestion");

  await ensureCollection();

  const files = findDocs(path.join(process.cwd(), "data"));


  for (const f of files) {
    await ingestFile(f);
  }

  console.log("🎉 DONE");
})();
